File: go/analyze/rules_stmtverb.go
// SPDX-License-Identifier: AGPL-3.0-or-later

package analyze

import (
	"regexp"
	"strings"
)

// -- Statement-verb helpers --------------------------------------------
//
// Many existing analyzer rules trip false positives because the regex
// matches keywords that appear LEXICALLY inside an unrelated SQL
// statement. Concrete cases observed in production (2026-05-12 example-service
// migration 234):
//
//   GRANT SELECT, INSERT, UPDATE, DELETE, TRUNCATE, REFERENCES, TRIGGER
//     ON TABLES TO example_service_app;
//
// trips:
//   no-truncate                — regex `\bTRUNCATE\b`
//   no-update-without-where    — regex `\bUPDATE\s+[^;]*;`
//
// even though both keywords are appearing as PostgreSQL privilege names
// in a GRANT clause, not as statement verbs.
//
// The proper fix is a real SQL parser (e.g. pg_query_go) but that's a
// heavy dep with cgo. The pragmatic fix used here is a statement-boundary
// + verb-gate helper: split the migration body into top-level statements
// (respecting `--` comments, `/* */` block comments, `$$tag$$ ... $$tag$$`
// dollar-quoted blocks, and string literals), then identify the leading
// keyword of each.
//
// False-positive-prone rules wrap their match with `stmtIsVerb(line,
// fileBody, "UPDATE")` so a TRUNCATE inside a GRANT only fires the
// no-truncate rule when the containing statement IS a TRUNCATE.

// reSingleLineComment matches a `--` line comment to the end of line.
var reSingleLineComment = regexp.MustCompile(`(?m)--[^\n]*`)

// reBlockComment matches `/* ... */` block comments (non-greedy).
var reBlockComment = regexp.MustCompile(`(?s)/\*.*?\*/`)

// reStringLiteral matches `'...'` literals, allowing `''` as an escaped
// single quote. Multi-line literals are allowed.
var reStringLiteral = regexp.MustCompile(`(?s)'(?:[^']|'')*'`)

// reVerb captures the leading SQL keyword of a statement (trimmed +
// upper-cased). PostgreSQL keywords are case-insensitive.
var reVerb = regexp.MustCompile(`(?i)^\s*([A-Z][A-Z_]+)\b`)

// stripCommentsAndLiterals erases the content of comments, dollar-quoted
// blocks, and string literals so subsequent statement-boundary detection
// doesn't trip on `;` characters inside them. The text length is
// preserved (replaced with spaces) so line / column offsets stay
// consistent with the original body.
//
// Dollar quotes (`$tag$ ... $tag$`) need tag-matching which RE2 can't
// express as a backreference, so we walk them by hand. Comments and
// single-quoted strings are simple regex masks.
func stripCommentsAndLiterals(body string) string {
	out := []byte(body)
	maskDollarQuotes(out, body)
	maskAll(out, reBlockComment, body)
	maskAll(out, reStringLiteral, body)
	maskAll(out, reSingleLineComment, body)
	return string(out)
}

func maskAll(dst []byte, re *regexp.Regexp, src string) {
	for _, loc := range re.FindAllStringIndex(src, -1) {
		for i := loc[0]; i < loc[1] && i < len(dst); i++ {
			// Preserve newlines so line counting stays correct.
			if dst[i] != '\n' {
				dst[i] = ' '
			}
		}
	}
}

// maskDollarQuotes finds `$tag$ ... $tag$` regions and replaces their
// content with spaces (preserving newlines). Tags can be empty (`$$`)
// or named (`$body$`). Properly tag-matched (no closure on a different
// tag).
func maskDollarQuotes(dst []byte, src string) {
	n := len(src)
	for i := 0; i < n; {
		if src[i] != '$' {
			i++
			continue
		}
		// Try to consume an opening tag: $tag$ where tag is
		// [a-zA-Z_][a-zA-Z0-9_]* (or empty).
		j := i + 1
		for j < n && (isAlnumUnder(src[j])) {
			j++
		}
		if j >= n || src[j] != '$' {
			i++
			continue
		}
		tag := src[i : j+1] // includes both $s
		// Find the matching close.
		close := strings.Index(src[j+1:], tag)
		if close < 0 {
			// Unterminated — treat as plain $ and move on.
			i++
			continue
		}
		end := j + 1 + close + len(tag)
		// Mask [i, end), preserving newlines.
		for k := i; k < end && k < len(dst); k++ {
			if dst[k] != '\n' {
				dst[k] = ' '
			}
		}
		i = end
	}
}

func isAlnumUnder(c byte) bool {
	return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') ||
		(c >= '0' && c <= '9') || c == '_'
}

// statementSlices returns the byte offsets of each top-level statement
// in body. Each returned [start, end) covers from the statement's first
// non-whitespace character through (and including) its terminating `;`.
// Top-level here means: outside of comments, string literals, and
// dollar-quoted blocks. We DON'T attempt to honor nested BEGIN…END
// blocks in PL/pgSQL — those are wrapped in dollar quotes in practice.
func statementSlices(body string) [][2]int {
	masked := stripCommentsAndLiterals(body)
	var out [][2]int
	start := -1
	for i, b := range []byte(masked) {
		if start == -1 {
			if b == ' ' || b == '\t' || b == '\n' || b == '\r' {
				continue
			}
			start = i
		}
		if b == ';' {
			out = append(out, [2]int{start, i + 1})
			start = -1
		}
	}
	// Trailing statement with no `;` (rare; pgroll emits these in some
	// strategies). Treat as one statement.
	if start != -1 && start < len(body) {
		out = append(out, [2]int{start, len(body)})
	}
	return out
}

// statementContaining returns the statement [start, end) of body that
// contains the byte offset `at`, or (-1, -1) when `at` is between
// statements (whitespace / between `;` and the next verb).
func statementContaining(body string, at int) (int, int) {
	for _, s := range statementSlices(body) {
		if at >= s[0] && at < s[1] {
			return s[0], s[1]
		}
	}
	return -1, -1
}

// statementVerb returns the upper-cased leading keyword of the statement
// containing `at` in body. Returns the empty string when the offset is
// between statements or when the statement doesn't start with a verb.
//
// Example callers (from rules_data.go / rules.go):
//
//	if statementVerb(f.Body, matchStart) != "UPDATE" { continue }
//	if statementVerb(f.Body, matchStart) != "TRUNCATE" { continue }
//
// This single gate eliminates the GRANT-clause false positives without
// touching the existing regex semantics for legitimate matches.
func statementVerb(body string, at int) string {
	a, b := statementContaining(body, at)
	if a < 0 {
		return ""
	}
	m := reVerb.FindStringSubmatch(body[a:b])
	if len(m) < 2 {
		return ""
	}
	return strings.ToUpper(m[1])
}

// stmtVerbIs is a convenience wrapper used by rule Check() bodies. It
// returns true when the statement containing the match's start offset
// has the given verb (case-insensitive).
func stmtVerbIs(body string, at int, verb string) bool {
	return statementVerb(body, at) == strings.ToUpper(verb)
}

File: rust/keystone-sdk/src/analyze/rules_stmtverb.rs
// SPDX-License-Identifier: Apache-2.0

//! Statement-verb helpers.
//!
//! Ported from the Go `rules_stmtverb.go`. Many rules trip false positives
//! because a keyword appears lexically inside an unrelated statement (e.g.
//! `TRUNCATE` as a privilege name in a `GRANT` clause). The fix is a
//! statement-boundary + leading-verb gate: split the body into top-level
//! statements (respecting `--`/`/* */` comments, `$tag$…$tag$` dollar
//! quotes, and `'…'` string literals), then identify each statement's
//! leading keyword. All offsets are BYTE offsets, matching Go and the
//! `regex` crate's `Match::start()`.

use std::sync::LazyLock;

use regex::Regex;

/// `--` line comment to end of line.
static RE_SINGLE_LINE_COMMENT: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"(?m)--[^\n]*").expect("single-line comment regex"));

/// `/* ... */` block comment (non-greedy, spans newlines).
static RE_BLOCK_COMMENT: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"(?s)/\*.*?\*/").expect("block comment regex"));

/// `'...'` literal allowing `''` as an escaped single quote (spans newlines).
static RE_STRING_LITERAL: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"(?s)'(?:[^']|'')*'").expect("string literal regex"));

/// Leading SQL keyword of a statement (case-insensitive).
static RE_VERB: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"(?i)^\s*([A-Z][A-Z_]+)\b").expect("verb regex"));

fn is_alnum_under(c: u8) -> bool {
    c.is_ascii_alphanumeric() || c == b'_'
}

/// Returns the first byte index of `needle` in `haystack`, or `None`.
fn find_subslice(haystack: &[u8], needle: &[u8]) -> Option<usize> {
    if needle.is_empty() {
        return Some(0);
    }
    if needle.len() > haystack.len() {
        return None;
    }
    haystack.windows(needle.len()).position(|w| w == needle)
}

/// Masks `$tag$ ... $tag$` regions (content → spaces, newlines preserved),
/// matching the close tag exactly. Tags may be empty (`$$`) or named.
fn mask_dollar_quotes(dst: &mut [u8], src: &str) {
    let s = src.as_bytes();
    let n = s.len();
    let mut i = 0;
    while i < n {
        if s[i] != b'$' {
            i += 1;
            continue;
        }
        let mut j = i + 1;
        while j < n && is_alnum_under(s[j]) {
            j += 1;
        }
        if j >= n || s[j] != b'$' {
            i += 1;
            continue;
        }
        let tag = &s[i..=j]; // includes both $s
        match find_subslice(&s[j + 1..], tag) {
            None => i += 1, // unterminated — treat as plain $
            Some(close) => {
                let end = j + 1 + close + tag.len();
                for k in i..end.min(dst.len()) {
                    if dst[k] != b'\n' {
                        dst[k] = b' ';
                    }
                }
                i = end;
            }
        }
    }
}

/// Masks all matches of `re` in `src` into `dst` (→ spaces, newlines kept).
fn mask_all(dst: &mut [u8], re: &Regex, src: &str) {
    for mat in re.find_iter(src) {
        for k in mat.start()..mat.end().min(dst.len()) {
            if dst[k] != b'\n' {
                dst[k] = b' ';
            }
        }
    }
}

/// Erases comments, dollar-quoted blocks, and string literals (→ spaces,
/// length + newlines preserved) so statement-boundary detection ignores
/// `;` inside them.
fn strip_comments_and_literals(body: &str) -> Vec<u8> {
    let mut out = body.as_bytes().to_vec();
    mask_dollar_quotes(&mut out, body);
    mask_all(&mut out, &RE_BLOCK_COMMENT, body);
    mask_all(&mut out, &RE_STRING_LITERAL, body);
    mask_all(&mut out, &RE_SINGLE_LINE_COMMENT, body);
    out
}

/// Byte offsets `[start, end)` of each top-level statement (first non-ws
/// char through its terminating `;`). A trailing statement without `;` is
/// returned as one slice.
fn statement_slices(body: &str) -> Vec<(usize, usize)> {
    let masked = strip_comments_and_literals(body);
    let mut out = Vec::new();
    let mut start: Option<usize> = None;
    for (i, &b) in masked.iter().enumerate() {
        if start.is_none() {
            if b == b' ' || b == b'\t' || b == b'\n' || b == b'\r' {
                continue;
            }
            start = Some(i);
        }
        if b == b';' {
            out.push((start.unwrap(), i + 1));
            start = None;
        }
    }
    if let Some(s) = start {
        if s < body.len() {
            out.push((s, body.len()));
        }
    }
    out
}

/// The statement `[start, end)` containing byte offset `at`, or `None`.
fn statement_containing(body: &str, at: usize) -> Option<(usize, usize)> {
    statement_slices(body)
        .into_iter()
        .find(|&(s, e)| at >= s && at < e)
}

/// The upper-cased leading keyword of the statement containing `at`, or the
/// empty string between statements / when no verb leads.
fn statement_verb(body: &str, at: usize) -> String {
    let Some((a, b)) = statement_containing(body, at) else {
        return String::new();
    };
    match RE_VERB.captures(&body[a..b]) {
        Some(c) => c
            .get(1)
            .map(|x| x.as_str().to_uppercase())
            .unwrap_or_default(),
        None => String::new(),
    }
}

/// True when the statement containing `at` has the given leading verb.
pub(crate) fn stmt_verb_is(body: &str, at: usize, verb: &str) -> bool {
    statement_verb(body, at) == verb.to_uppercase()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn verb_gate_ignores_grant_clause() {
        let body = "GRANT SELECT, INSERT, UPDATE, DELETE, TRUNCATE ON TABLE t TO app;";
        // The TRUNCATE keyword is inside a GRANT statement.
        let at = body.find("TRUNCATE").unwrap();
        assert!(stmt_verb_is(body, at, "GRANT"));
        assert!(!stmt_verb_is(body, at, "TRUNCATE"));
    }

    #[test]
    fn verb_gate_matches_real_truncate() {
        let body = "TRUNCATE TABLE t;";
        let at = body.find("TRUNCATE").unwrap();
        assert!(stmt_verb_is(body, at, "TRUNCATE"));
    }

    #[test]
    fn semicolon_in_string_literal_does_not_split() {
        let body = "INSERT INTO t VALUES ('a;b'); TRUNCATE t;";
        let at = body.find("TRUNCATE").unwrap();
        assert!(stmt_verb_is(body, at, "TRUNCATE"));
    }

    #[test]
    fn dollar_quote_block_is_masked() {
        let body = "DO $$ BEGIN TRUNCATE t; END $$; SELECT 1;";
        // The TRUNCATE inside the DO block belongs to the DO statement.
        let at = body.find("TRUNCATE").unwrap();
        assert!(stmt_verb_is(body, at, "DO"));
    }
}
